### PR TITLE
[RESUBMIT] Standardize on error types for distributed errors.

### PR DESCRIPTION
**Pull request by pritamdamania87**
We have a plethora of error types for various errors raised from c10d. These include `RuntimeError`, `TimeoutError`, `SocketError`, `DistBackendError` etc.

This results in messy code during error handling somewhat like this:
```
if "NCCL" in exception_str:
  ...
if "Timed out initializing process group in store based barrier on rank" in exception_str:
  ...
if "The client socket has timed out after" in exception_str:
  ...
if "Broken pipe" in exception_str:
  ...
if "Connection reset by peer" in exception_str:
  ...
```

To address this issue, in this PR I've ensured added these error types:

1. **DistError** - the base type of all distributed errors
2. **DistBackendError** - this already existed and referred to PG backend errors
3. **DistStoreError** - for errors originating from the store
4. **DistNetworkError** - for general network errors coming from the socket library

### CODE REVIEW

**Comment by pritamdamania87** [comment]
Resubmit of https://github.com/pytorch/pytorch/pull/107651, only change is that I used `RuntimeError` for errors like `Distributed package doesn't have NCCL built in`. This should fix failures like https://github.com/pytorch/pytorch/actions/runs/6005354009/job/16289079518 and https://github.com/pytorch/pytorch/actions/runs/6005354009/job/16289079928

**Comment by pritamdamania87** [comment]
@H-Huang Can we add all the appropriate labels to this PR to ensure all tests are run (including inductor ones)? :)

**Comment by pritamdamania87** [comment]
One more change from https://github.com/pytorch/pytorch/pull/107651 is that I made sure DistError is a base class of DistStoreError, DistNetworkError and DistBackendError in python land too.

**Review by H-Huang — COMMENTED** [review]
Resubmit looks good, will wait for CI

File: inductor_graph_breaks.csv [hunk 1]
```diff
@@ -0,0 +1,5 @@
+model,reason,user_stack
```
Comment (H-Huang) [inline inductor_graph_breaks.csv:1]: is this supposed to be checked in?

Comment (pritamdamania87) [reply inductor_graph_breaks.csv:1]: Good catch, no this was unintentional.

**Review by H-Huang — APPROVED** [review]
lgtm

**Pull request closed without merging**